util-images/network/netperfbenchmark/pkg/worker/util.go

```diff
@@ -0,0 +1,204 @@
+limitations under the License.
+*/
+
+package worker
+
+import (
+	"errors"
+	"math"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"k8s.io/klog"
+)
+
+// Protocols supported
+const (
+	ProtocolTCP  = "TCP"
+	ProtocolUDP  = "UDP"
+	ProtocolHTTP = "HTTP"
+)
+
+// Number of metrics for each protocol
+const (
+	tcpMetricsCount  = 2
+	udpMetricsCount  = 11
+	httpMetricsCount = 12
+)
+
+// Iperf results vary from protocol,required for proper parsing
+const (
+	zeroFormatTCP         = "0.0"
+	fractionZeroFormatTCP = ".0"
+	zeroFormatUDP         = "0.00"
+	fractionZeroFormatUDP = ".00"
+)
+
+// Function to be applied for each metric for aggregation
+var (
+	iperfUDPFunction = []string{"", "", "", "Sum", "Sum", "Sum", "Sum", "Sum", "Avg", "Avg", "Min", "Max", "Avg", "Sum"}
+	iperfTCPFunction = []string{"", "", "", "Sum", "Avg"}
+)
+
+// parseResult parses the response received for each protocol type
+func parseResult(protocol string, result []string, testDuration string) ([]float64, error) {
+	switch protocol {
+	case ProtocolTCP:
+		return parseIperfResponse(result, testDuration, ProtocolTCP, tcpMetricsCount, zeroFormatTCP, fractionZeroFormatTCP, iperfTCPFunction)
+	case ProtocolUDP:
+		return parseIperfResponse(result, testDuration, ProtocolUDP, udpMetricsCount, zeroFormatUDP, fractionZeroFormatUDP, iperfUDPFunction)
+	case ProtocolHTTP:
+		return parseSiegeResponse(result)
+	default:
+		return nil, errors.New("invalid protocol: " + protocol)
+	}
+}
+
+func parseIperfResponse(result []string, testDuration string, protocol string, metricCount int, zeroFormat string, fractionZeroFormat string, operators []string) ([]float64, error) {
+	aggregatedResult := make([]float64, 0, metricCount)
+	unitRegex := regexp.MustCompile(`%|\[\s+|\]\s+|KBytes\s+|KBytes/sec\s*|sec\s+|pps\s*|ms\s+|/|\(|\)\s+`)
+	multiSpaceRegex := regexp.MustCompile(`\s+`)
+	hyphenSpaceRegex := regexp.MustCompile(`\-\s+`)
+	count := 0
+	sessionID := make(map[string]bool)
+	for _, row := range result {
+		formattedString := hyphenSpaceRegex.ReplaceAllString(multiSpaceRegex.ReplaceAllString(unitRegex.ReplaceAllString(row, " "), " "), "-")
+		if !strings.Contains(formattedString, zeroFormat+"-"+testDuration+fractionZeroFormat) {
+			continue
+		}
+		split := strings.Split(formattedString, " ")
+		// iperf gives aggregated result in a row with "SUM", but many fields are not getting aggregated
+		// hence not depending on iperf's aggregation
+		if len(split) < metricCount+1 || "SUM" == split[1] || split[2] != zeroFormat+"-"+testDuration+fractionZeroFormat {
+			continue
+		}
+		// iperf sometimes prints duplicate rows for same session id(for same duration), making sure
+		// the row is taken for calculation only once.
+		if _, ok := sessionID[split[1]]; ok {
+			continue
+		}
+		for i, v := range split {
+			if i == 1 {
+				sessionID[v] = true
+				continue
+			}
+			//first index and hte last is ""
+			if i == 0 || i == 2 || i == metricCount+3 {
```

Comment:
given this and the above, it would clear to have:
```
for i := 3; i < len(split); i++ {
  ...
}
```

Also - how many items in split can be? Can there be more than "metricsCount+3"?

Reply:
Modified the loop to avoid specific checks (metricsCount+3 would be the last element in split).